tests/test_filters.py

```diff
@@ -1229,19 +1229,57 @@ def test_extract_surface():
     assert surf_no_subdivide.n_faces == 6
 
 
-def test_merge_general():
-    mesh = examples.load_uniform()
-    thresh = mesh.threshold_percent([0.2, 0.5], progress_bar=True)  # unstructured grid
-    con = mesh.contour()  # poly data
+def test_merge_general(uniform):
+    thresh = uniform.threshold_percent([0.2, 0.5], progress_bar=True)  # unstructured grid
+    con = uniform.contour()  # poly data
     merged = thresh + con
     assert isinstance(merged, pyvista.UnstructuredGrid)
     merged = con + thresh
     assert isinstance(merged, pyvista.UnstructuredGrid)
     # Pure PolyData inputs should yield poly data output
-    merged = mesh.extract_surface() + con
+    merged = uniform.extract_surface() + con
     assert isinstance(merged, pyvista.PolyData)
 
 
+def test_iadd_general(uniform, hexbeam, sphere):
+    unstructured = hexbeam
+    sphere_shifted = sphere.copy()
+    sphere_shifted.points += [1, 1, 1]
+    # successful case: poly += poly
+    merged = sphere
+    merged += sphere_shifted
+    assert merged is sphere
+
+    # successful case: unstructured += anything
+    merged = unstructured
+    merged += uniform
+    assert merged is unstructured
+    merged += unstructured
+    assert merged is unstructured
+    merged += sphere
+    assert merged is unstructured
+
+    # failing case: poly += non-poly
+    merged = sphere
+    merged += uniform
+    assert merged is not sphere
```

Comment:
If the merge fails, we need to raise error message or a warning message.

Reply:
As it is designed now, `+=` never fails. `merge(inplace=True)` can fail, and that raises as it should. I tried to discuss this in the PR comment.

Reply:
I am +1 for change the design to raise error. How about @pyvista/developers ?

Reply:
If we add this, Id vote for raising an error.  By adding it, we are specifically indicating that in place operations are supported.  Then the error message is useful as documentation on what should be done.

Reply:
@MatthewFlamm the reason why I'm not so sure about that is because this works:
```py
tup = (1, 2)
tup += (3, 4)
```
even though the operation is not in-place. The reason why I'm a bit on the fence is because `type(self)` in itself doesn't tell you if it's inplace.

I can accept both approaches. I'm curious what @akaszynski thinks.

Reply:
The distinction I see is that tuples _always_ return a new object.  Here we would _sometimes_ return a new object with a different type.  Having said that, I don't have a strong feeling about it.

Reply:
### Mutable vs. Immutable `__iadd__`
Perhaps this isn't intuitive, but Python implements `__iadd__` for its immutable types by returning a new instance of that type. Most notable is `int`:
```py
# int

>>> my_int = 10
>>> id(my_int)
9789280
>>> my_int += 100
>>> id(my_int)
9792480
```

The same occurs with `float`, `str`, `bool`, and `tuple`. I think this is done out of convenience rather than efficiency. For example, rather than writing:
```py
my_string = 'hello'
my_string = my_string + 'world'
```
It's a little easier to write:
```py
my_string = 'hello'
my_string += 'world'
```
In both cases, the memory address of `my_string` changes after assignment, so it's following the "return new instance if not mutable" strategy.

It should be noted that `__add__` and `__iadd__` for these immutable types take the same amount of time to execute.

Mutable types however modify the object in situ without changing the memory address of the parameter:
```py
>>> my_list0 = ['a', 1]
>>> id(my_list0)
140397782211072
>>> my_list1 = ['b', 2]
>>> my_list0 += ['a', 1]
>>> id(my_list0)
140397782211072
```
---
### How we should Implement this in PyVista

The difference between the base Python types and our PyVista types is may return a different type (though the same base type). This makes it more unclear how to operate because the user might be expecting to be able to add two `RectlinearGrid` objects together and return the same type, even if the memory address of the original reference has changed.

Since our data types are mutable, and I've been unable to find a python example that adds two different types in situ, I think we should target the following behavior
```py
     Unif Rect Unst Poly Stru
Unif ['e', 'e', 'e', 'e', 'e']
Rect ['e', 'e', 'e', 'e', 'e']
Unst ['m', 'm', 'm', 'm', 'm']
Poly ['e', 'e', 'e', 'm', 'e']
Stru ['e', 'e', 'e', 'e', 'e']
```
This way it's clear that when we `__iadd__`, the type of the object we're "mutating" does not change. If `Rect + Rect` could return the same type, even though it's a new instance, I think that would work. However, in our current implementation `Rect + Rect -> Unst`, and this type change is not reflected in Python.

Reply:
OK, thank you all for weighing in :)